### ISSUE TITLE
issue with static data structures in rucio output

### ISSUE THREAD

**Issue by vkuznet** [rucio/rucio#1752]
Motivation
----------
While working with Rucio data records I found that their structure is not static in terms of data attributes. By that I mean that returned dictionary contains dynamic set of keys, plus it mixed pre-defined attributes names with attribute values. To demonstrate this I'll take one particular record from replicas API, e.g.
```
 curl -H "Accept: application/json" -H "X-Rucio-Auth-Token: $token" http://cms-rucio-test.cern.ch/replicas/cms/Charmonium/Run2017D-31Mar2018-v1/MINIAOD#56f1f7ec-3789-11e8-a513-02163e018570
```
This API returns the following records
```
{"states": {"T2_US_Nebraska": "AVAILABLE"}, "pfns": {}, "adler32": "06273a48", "name": "/store/data/Run2017D/Charmonium/MINIAOD/31Mar2018-v1/90000/F8AD7081-A037-E811-BF64-0CC47AC52BAE.root", "rses": {"T2_US_Nebraska": []}, "scope": "cms", "bytes": 2877629555, "md5": null}
{"states": {"T2_US_Nebraska": "AVAILABLE"}, "pfns": {}, "adler32": "881ec8cf", "name": "/store/data/Run2017D/Charmonium/MINIAOD/31Mar2018-v1/90000/FA607615-8E37-E811-B2DE-0CC47AC52C8E.root", "rses": {}, "scope": "cms", "bytes": 3890770828, "md5": null}
```

These records contain static attribute names, such as `states`, `md5`, `scope`, etc. and within `states` part of the records they contain dynamic keys, in this case both records return `T2_US_Nebraska`, but in general it would be name of the rse. As you can see, the record keys (attributes) will be different in different type of queries since you used `rse` values as record key (attribute) of states data-structure. This is represent certain problem to treat these records in generic way, e.g. I can't map a record into static data-structure since records keys will be dynamic.

Instead, rucio should return static data-strictures all the times. For instance the states sub-structure may be represented as following:
```
"states":{"rse":"T2_US_Nebraska, "state":"AVAILABLE"}
or to be generic for multiple rses as
"states":[{"rse":"T2_US_Nebraska, "state":"AVAILABLE"}, {"rse":"T2_XXX", "state":"STATE_OF_RSE"}]
```
In this case the states has static data-structure where all keys (attributes) are well defined, i.e. in this example there are two keys: `rse` and `state`, and such structures will remain static for infinitive number of RSE/STATE values.

I wonder how many APIs inherit this behavior and how hard/easy will be to fix them. I also need to know how many of such dynamic structures exists in Rucio APIs output.

Another issue with such dynamic data-structures affects long-term analytics (if we plan to do it) where records keys will grow over time and will not represent finite set of attributes.

For the record, this issue is very critical for CMS Data Aggregation System which will consume this data and need to aggregate it with rest of CMS data-service records. In order to do aggregation several issues should be solve, including this particular one.

I'm CCing @ericvaandering Eric to follow up this issue. Eric, Rucio falls into the same trap as many other CMS data-services by using key/attribute values as keys of the underlying python dict keys. This represent issue with developing generic parser of Rucio records.

**Comment by bari12** [comment]
I am not quite sure why you don't transform the result into a data structure you can store? This should not be very resource intensive and you can guarantee this way that you store sanitised data.

This pattern of key/attribute values is certainly used in more APIs, I would have to check which ones though.
We cannot change the result of this API due to backwards compatibility with old clients. We certainly could make a v2 API, or some kind of option to return the result in this kind of way.

**Comment by vkuznet** [comment]
Martin,
the transformation implies full knowledge of all keys. Since it is not possible
in this structure such process can be very tedious (not only on code writing)
but on parsing level too. For instance, how many keys exists, will these keys
stay the same over long time (I don't think so since we may add discard RSEs for
instance), how many structure have this issue in all Rucio APIs, what is
complete schema of all APIs. The list can go on. Not to mention that
custom transformation cost time and resources.

I outlined the issue and presented use-cases. Anyone/tool who need to deal
with aggregation or analytics of the results will face this issue. When you need
to deal with dozens of services in real time it is a big issue.

But of course I understand backward compatibility issue and ideally it should be
fixed both on server and client sides. If is it feasible it is another story.

Of course I can work around such cases, but I'm in favor that this issue should
be fixed eventually.

**Comment by bbockelm** [comment]
Hi Martin,

Aren’t at least a few of the APIs passed through a JSON schema verification?  Or maybe I’m just recalling some internal data structures?

It’d probably be useful to maintain some sort of published schema as new users of the API come along.  Success breeds these sort of questions, after all!

Thanks,

Brian

**Comment by bari12** [comment]
We only do the schema verification for the incoming data structures (https://github.com/rucio/rucio/blob/master/lib/rucio/common/schema/atlas.py) not for the outgoing one. But I fully agree, we should maybe have something for the return structures too. Not so much a verification, but more documentation.
@vkuznet I get your point, I didn't mean to do the transformation for all eternity, just for now; The data structures right now are very much prepared in a way so that the client (The rucio client that is) can immediately deal with them efficiently. (That's why the rses are keys, so the client can quickly get all replica states for a certain rse in a long list of replicas) - I think what might be most appropriate is an option for the call to prepare the return object in an "analytics ready" way. We would have to look at all REST calls you are making and see if such an option is needed and keep it in mind if there are API changes in the future.

**Comment by vkuznet** [comment]
Martin it is not a stopper for me since I deal with these issues in other data-services, but eventually I'll be glad if this issue will be fixed one way or another.

**Comment by vkuznet** [comment]
So far I identified the following dynamic structures in /replicas/scope/name API:
- states
```
"states":{"T2_US_Nebraska":"AVAILABLE"}}
```
- rses
```
"rses":{"T2_US_Nebraska":["gsiftp://red-gridftp.unl.edu/user/uscms01/pnfs/unl.edu/data4/cms//store/data/Run2017D/Charmonium/MINIAOD/31Mar2018-v1/90000/FA607615-8E37-E811-B2DE-0CC47AC52C8E.root"]}
```
- pfns
```
"pfns":{"gsiftp://red-gridftp.unl.edu/user/uscms01/pnfs/unl.edu/data4/cms//store/data/Run2017D/Charmonium/MINIAOD/31Mar2018-v1/90000/FA607615-8E37-E811-B2DE-0CC47AC52C8E.root":{"client_extract":false,"domain":"wan","priority":1,"rse":"T2_US_Nebraska","type":"DISK","volatile":false}}
```

**Comment by vkuznet** [comment]
The pfns example is extremely bad if such records will be placed in DB and indexed on these keys since (a) pfn names are long strings, (b) we have extremely long list of pfns in GRID universe.

**Comment by mlassnig** [comment]
Hi Valentin, I'm not really following you. What is your concern with the pfns? They are not stored in the database at all?

**Comment by vkuznet** [comment]
Yes, but if we will dump data to HDFS (as is) and then create a dataframe the
pfns will become attributes (column names) rather then values.

**Comment by bbockelm** [comment]
Hi @vkuznet -

I think we've determined that this ticket is really about having a JSON schema for the Rucio APIs, no?  Since that comes down to a pretty major redesign of the APIs (which, FWIW, is being discussed), what should we do with this ticket?

Brian

**Comment by vkuznet** [comment]
I think it is up to Rucio team. We can use it for progress or further examples (as I did today), or it can be turned into actions/milestones.

**Comment by bari12** [comment]
Discussed in the development meeting on [2018-11-08](https://indico.cern.ch/event/766990/). We decided to leave it like this until we have a V2 API (Long term project)

**Comment by ericvaandering** [comment]
Should this become an real problem for wanting to dump data to HDFS, it should be possible as an interim step to define the desired schema(s) for v2 and build a simple translator which would precede v2, right?

**Comment by bari12** [comment]
yes, that's definitely an option.
